datastore/gcloud/aio/datastore/utils.py

```diff
@@ -1,12 +1,25 @@
+from datetime import datetime as dt
 from typing import Any
 from typing import Dict
 
 from gcloud.aio.datastore.constants import FORMATTERS
 from gcloud.aio.datastore.constants import TypeName
-from gcloud.aio.datastore.constants import TYPES
 from gcloud.aio.datastore.constants import UNFORMATTERS
 
 
+# TODO: add types for entityValue, geoPointValue and arrayValue
+# NOTE: TypeName.KEY is added dynamically by the datastore
```

Comment:
I think I'm missing the reason behind all this metaprogramming and adding the typename for this constant. Let me know where I'm misunderstanding this requirement:
- we want to support `Value.from_repr()` and `Value.to_repr()` for `keyValue` types

Is this not just plumbing? ie. in `Value.from_repr()` we'd check if the entity is a `keyValue` type, and if so we'd set: `self.value = Key.from_repr(...)`. Then, in `Value.to_repr()` we'd emit `self.value.to_repr()` if it is a subclass/instanceof `Key`. I don't think we need to modify the `TYPES` dictionary dynamically at all, we just need to add `Key: TypeName.KEY` to it.

Reply:
The problem is that you can subclass the datastore and have a key type that's completely unrelated to Key.

I can do something like this:
```
from gcloud.aio.datastore import Datastore

class MyKey:
  pass

class MyDatastore(Datastore):
  key_kind = MyKey
```
But when the client tries to run a query with a filter on a value of type MyKey, it fails inside `Value.to_repr()` because MyKey is an unknown type.

That's why I need Value to know about the datastore's key_kind, which has been kind of a headache. But I like this implementation - the TYPES dict is basically known at the creation time of your custom datastore.

Reply:
Makes sense! Let's try to solve it the same way as we've done elsewhere -- by adding `key_kind` (and `entity_kind`, if we wanted to support both`) as class attributes to `Value`, like we do for all the other classes.

Then, within and `Value` methods, we'd always reference `self.key_kind` instead of `Key`. The `Entity` and `EntityResult` classes have good examples of how to do this.

Reply:
I like the idea.
I would still like infer_type, make_value and parse_value to work with `key_kind` though. At least make_value, since that would add support for `key_kind` to GQL queries.

The problem is that there is some duplication between Value's form_repr/to_repr and make_value/perse_value. We can probably find a way to use only one of them.